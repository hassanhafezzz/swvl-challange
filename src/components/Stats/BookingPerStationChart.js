import React, { useContext } from 'react';
import { Bar } from 'react-chartjs-2';
import classNames from 'classnames/bind';
import { Context } from '../../store';
import styles from './styles.module.css';

const cx = classNames.bind(styles);

const BookingPerStationChart = () => {
  const [state] = useContext(Context);
  const { bookings, stations } = state;

  const stationsName = stations.map((station) => station.name);
  const counts = [];

  stationsName.forEach((station) => {
    let count = 0;
    bookings.forEach((booking) => {
      if (booking.pickupStation === station) {
        count += 1;
      }
    });
    counts.push(count);
  });

  const data = {
    labels: stationsName,
    datasets: [
      {
        label: 'bookings per station',
        data: counts,
        barThickness: 12,
        barPercentage: 0.5,
        backgroundColor: '#ff006e',
        borderColor: '#ff006e',
        borderWidth: 1,
      },
    ],
  };

  const options = {
    scales: {
      xAxes: [
        {
          ticks: {
            fontColor: '#222',
          },
          gridLines: {
            color: 'rgba(0, 0, 0, 0)',
          },
        },
      ],
      yAxes: [
        {
          ticks: {
            fontStyle: 'bold',
            precision: 0,
          },
          gridLines: {
            color: 'rgba(0, 0, 0, 0)',
          },
        },
      ],
    },
  };

  return (
    <section className={cx('section')} data-testid="booking-per-station-chart">
      <h4 className={cx('section-header')}>
        Break down for customers&apos; booking per station
      </h4>
      <Bar data={data} options={options} />
    </section>
  );
};

export default BookingPerStationChart;
