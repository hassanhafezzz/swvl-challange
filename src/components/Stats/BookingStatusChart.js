import React, { useContext } from 'react';
import { Doughnut } from 'react-chartjs-2';
import classNames from 'classnames/bind';
import { BOOKING_STATUS } from '../../constants';
import { Context } from '../../store';
import styles from './styles.module.css';

const cx = classNames.bind(styles);

const BookingStatusChart = () => {
  const [state] = useContext(Context);
  const { bookings } = state;

  const { COMPLETED, MISSED, CANCELLED } = BOOKING_STATUS;
  const bookingData = { [COMPLETED]: 0, [MISSED]: 0, [CANCELLED]: 0 };

  bookings.forEach((booking) => {
    const { status } = booking;
    bookingData[status] += 1;
  });

  const labels = Object.keys(bookingData);
  const values = Object.values(bookingData);

  const data = {
    labels,
    datasets: [
      {
        label: 'Booking Status',
        data: values,
        backgroundColor: ['#00c785', '#ffbe0b', '#ff006e'],
        borderColor: ['#00c785', '#ffbe0b', '#ff006e'],
        borderWidth: 1,
      },
    ],
  };

  return (
    <section className={cx('section')} data-testid="booking-status-chart">
      <h4 className={cx('section-header')}>
        Break down for customers&apos; status for the trip
      </h4>
      <Doughnut data={data} />
    </section>
  );
};

export default BookingStatusChart;
