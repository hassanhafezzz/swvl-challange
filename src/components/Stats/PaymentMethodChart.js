import React, { useContext } from 'react';
import { Pie } from 'react-chartjs-2';
import classNames from 'classnames/bind';
import { PAYMENT_METHODS } from '../../constants';
import { Context } from '../../store';
import styles from './styles.module.css';

const cx = classNames.bind(styles);

const BookingStatusChart = () => {
  const [state] = useContext(Context);
  const { bookings } = state;

  const labels = [PAYMENT_METHODS.CASH, PAYMENT_METHODS.CREDIT];
  const values = [0, 0];

  bookings.forEach((booking) => {
    const { paymentMethod } = booking;
    if (paymentMethod === PAYMENT_METHODS.CASH) {
      values[0] += 1;
    } else {
      values[1] += 1;
    }
  });

  const data = {
    labels,
    datasets: [
      {
        label: 'Payment methods',
        data: values,
        backgroundColor: ['#8300d5', '#ff006e'],
        borderColor: ['#8300d5', '#ff006e'],
        borderWidth: 1,
      },
    ],
  };

  return (
    <section className={cx('section')}>
      <h4 className={cx('section-header')}>
        Break down for customers&apos; payment methods
      </h4>
      <Pie data={data} />
    </section>
  );
};

export default BookingStatusChart;
