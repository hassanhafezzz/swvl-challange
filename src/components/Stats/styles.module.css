@value utils: '../../styles/utils.css';

.container {
  composes: container from utils;
}

.heading {
  margin: 15px 0 10px;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.title {
  margin: 5px 0 10px;
  font-size: var(--h1-size);
}

.wrapper {
  margin: 50px auto;
  max-width: 600px;
}

.section {
  text-align: center;
}

.section-header {
  font-weight: var(--bold);
  font-size: var(--h3-size);
}

.divider {
  composes: divider from utils;
}

.divider.large {
  margin: 60px 0;
}
