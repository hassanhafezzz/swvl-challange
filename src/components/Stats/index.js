import React from 'react';
import classNames from 'classnames/bind';
import BookingPerStationChart from './BookingPerStationChart';
import BookingStatusChart from './BookingStatusChart';
import PaymentMethodChart from './PaymentMethodChart';
import Button, { BUTTON_VARIANT } from '../common/Button';
import styles from './styles.module.css';

const cx = classNames.bind(styles);

export default function Stats() {
  return (
    <div className={cx('container')} data-testid="stats">
      <div className={cx('heading')}>
        <div>
          <h3 className={cx('title')}>
            Trip Statistics{' '}
            <span role="img" aria-label="stats">
              ðŸ“Š
            </span>
          </h3>
        </div>
        <Button variant={BUTTON_VARIANT.SECONDARY} to="/">
          go back
        </Button>
      </div>

      <hr className={cx('divider')} />
      <div className={cx('wrapper')}>
        <BookingPerStationChart />
        <hr className={cx('divider', 'large')} />
        <BookingStatusChart />
        <hr className={cx('divider', 'large')} />
        <PaymentMethodChart />
      </div>
    </div>
  );
}
