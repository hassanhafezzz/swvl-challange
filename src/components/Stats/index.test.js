import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render } from '@testing-library/react';
import StoreProvider from '../../store';
import Stats from './index';

const ele = (
  <StoreProvider>
    <Router>
      <Stats />
    </Router>
  </StoreProvider>
);

describe('stats component test', () => {
  test('renders stats', () => {
    const { getByTestId } = render(ele);
    const stats = getByTestId('stats');
    expect(stats).toBeTruthy();
  });

  test('renders three charts', () => {
    const { getByTestId } = render(ele);
    const bookingPerStationChart = getByTestId('booking-per-station-chart');
    const bookingStatusChart = getByTestId('booking-status-chart');
    const paymentChart = getByTestId('payment-chart');

    expect(bookingPerStationChart).toBeTruthy();
    expect(bookingStatusChart).toBeTruthy();
    expect(paymentChart).toBeTruthy();
  });

  test('add snapshot for stats', () => {
    const { container, asFragment } = render(ele);
    expect(asFragment(container)).toMatchSnapshot();
  });
});
