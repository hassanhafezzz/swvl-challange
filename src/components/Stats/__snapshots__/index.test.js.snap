// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stats component test add snapshot for stats 1`] = `
<DocumentFragment>
  <div
    class="container"
    data-testid="stats"
  >
    <div
      class="heading"
    >
      <div>
        <h3
          class="title"
        >
          Trip Statistics 
          <span
            aria-label="stats"
            role="img"
          >
            ðŸ“Š
          </span>
        </h3>
      </div>
      <a
        class="button secondary"
        href="/"
      >
        go back
      </a>
    </div>
    <hr
      class="divider"
    />
    <div
      class="wrapper"
    >
      <section
        class="section"
        data-testid="booking-per-station-chart"
      >
        <h4
          class="section-header"
        >
          Break down for customers' booking per station
        </h4>
        <canvas
          height="150"
          width="300"
        />
      </section>
      <hr
        class="divider large"
      />
      <section
        class="section"
        data-testid="booking-status-chart"
      >
        <h4
          class="section-header"
        >
          Break down for customers' status for the trip
        </h4>
        <canvas
          height="150"
          width="300"
        />
      </section>
      <hr
        class="divider large"
      />
      <section
        class="section"
        data-testid="payment-chart"
      >
        <h4
          class="section-header"
        >
          Break down for customers' payment methods
        </h4>
        <canvas
          height="150"
          width="300"
        />
      </section>
    </div>
  </div>
</DocumentFragment>
`;
